OTP-19 and openssl 1.0.x compatibility

mworrell opened the pull request:
This changes a couple of things:

 * Sets min OTP version to 19, as it does indeed work excellent with OTP-19
 * Uses the openssl 1.0.x code to add the SANs to the certificate, as quite some machines are still using a late version of openssl 1.0.x
 * Enables the Dialyzer Makefile target
 * Small start with additional typing of arguments

gbour commented:
Hi @mworrell 
Thanks for your valuable contributions. I will review this MR quickly

gbour commented on Makefile:
```diff
@@ -5,8 +5,6 @@ all:
-	if [[ "$$TRAVIS_OTP_RELEASE" > "18" ]] ; then\
-		./rebar3 dialyzer;\
-	fi
+	./rebar3 dialyzer;\
```
```suggestion
	./rebar3 dialyzer;
```

gbour commented on src/letsencrypt_ssl.erl:
```diff
@@ -18,25 +18,22 @@
     private_key({new, "letsencrypt.key"}, CertsPath);
 
 private_key({new, KeyFile}, CertsPath) ->
     FileName = CertsPath++"/"++KeyFile,
```
this does not works well if either `CertsPath`or `KeyFile` or both are binary strings

gbour commented on src/letsencrypt_ssl.erl:
```diff
@@ -47,18 +44,14 @@ private_key(KeyFile, _) ->
--spec cert_request(string(), string(), list(string())) -> letsencrypt:ssl_csr().
+-spec cert_request(letsencrypt:domain(), file:filename_all(), list(letsencrypt:domain())) -> letsencrypt:ssl_csr().
 cert_request(Domain, CertsPath, SANs) ->
     KeyFile  = CertsPath ++ "/" ++ Domain ++ ".key",
```
same comment as previous one

mworrell replied:
Indeed, fixing.

mworrell commented:
Merged fixes.

mworrell commented:
@gbour could you have another look?

I also added:

 - Support for SANs using _slave_ mode
 - Replace Jiffy with JSX to fix problems with compilation on OTP-23
 - Renamed `letsencrypt_api:order/4` and `order/5` to `get_order/4` and `new_order/5`

mworrell commented:
Hi @gbour. GitHub doesn't show me clearly what changes you requested are still open.  Could you have a look?

mworrell commented:
Hi @gbour, do you have time to have a look at this?

mworrell commented:
Ping :)

Olivier-Boudeville commented:
Hi Marc, thanks for these improvements that, as I understand, could not be (yet) integrated here (FWIW, I took inspiration from them for https://github.com/Olivier-Boudeville/Ceylan-LEEC - at least for the SAN certificates); thanks to you and Guillaume for bringing ACME support to the Erlang community!

mworrell commented:
Hi @Olivier-Boudeville, thanks for the nice words, and good to know that this code is of good use for others as well. As for the pending status of this request, I guess that @gbour is very busy with other things at the moment.
